subprojects/file-collections/src/integTest/groovy/org/gradle/api/file/ConfigurableFileCollectionAssignmentIntegrationTest.groovy

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.file
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+
+class ConfigurableFileCollectionAssignmentIntegrationTest extends AbstractIntegrationSpec {
+
+    def "test Groovy FileCollection types assignment for #description"() {
+        given:
+        def buildFileDefinition = """
+            abstract class MyTask extends DefaultTask {
+                @Internal
+                {inputDeclaration}
+                @TaskAction
+                void run() {
+                    println("Result: " + input.files.collect { it.name })
+                }
+            }
+            tasks.register("myTask", MyTask) {
+                input $operation {inputValue}
+            }
+        """
+
+        when:
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "$eagerInputType input = project.files() as $eagerInputType")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        run("myTask")
+        outputContains(eagerResult)
+
+        when:
+        buildFile.text = buildFileDefinition
+            .replace("{inputDeclaration}", "abstract $lazyInputType getInput()")
+            .replace("{inputValue}", inputValue)
+
+        then:
+        run("myTask")
+        outputContains(lazyResult)
+
+        where:
+        description                       | operation | eagerInputType   | lazyInputType                | inputValue                         | eagerResult       | lazyResult
+        "FileCollection = FileCollection" | "="       | "FileCollection" | "ConfigurableFileCollection" | 'files("a.txt")'                   | 'Result: [a.txt]' | 'Result: [a.txt]'
```

Comment:
🤔 How about removing `Result: ` from here and adding the prefix in the assertion only? To shorten the lines.